runtime/memprof.c

```diff
@@ -189,6 +192,16 @@ CAMLprim value caml_memprof_set(value v)
   if(sz < 0 || !(l >= 0.) || l > 1.)
     caml_invalid_argument("caml_memprof_set");
 
+  /* This call to [caml_memprof_set] may stop sampling or change the
+     callback. We have to make sure that the postponed queue is empty
+     before continuing. */
+  if(!caml_memprof_suspended)
+    caml_memprof_handle_postponed();
+  else
+    /* But if we are currently running a callback, there is nothing
+       else we can do than purging the queue. */
+    purge_postponed_queue();
```

Comment:
Why is it the right choice to purge the queue here? This means that some blocks will be forgotten about, with no callback called on them. We could also decide to call the postponed blocks with the new callback, which is not great but sounds better to me that never calling any callback on them.

(I thought about keeping a copy of the current callback value in the postponed list, but I think that would interact badly with resources: in theory I could have a callback that looks into an open file descriptor, and close that descriptor as I change the callback, and future calls to the old callback would fail in odd ways. On the other hand, we could also ask users to do this kind of cleanup in a finalizer on the callback, or have the installing-the-callback function also take a cleanup function as parameter, so that we can call it at the right time.)

If the user stops the sampling in the current run, maybe it would be fine to still handle the postponed queue when the sampling has been stopped? (For example by calling `caml_memprof_handle_postponed();` again at the end of this function.)

Reply:
The handler used when sampling is stopped is `assert false`, which means we cannot call it after stopping.